File: Algorithm_Rust/examples/143. 重排链表.ts
class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

function reorderList(head: ListNode | null): void {
  if (!head) return;
  let arr: ListNode[] = [];
  let node: ListNode | null = head;
  while (node) {
    arr.push(node);
    node = node.next;
  }
  let count = 0;
  let pre_node = null as ListNode | null;
  while (count < Math.floor(arr.length / 2)) {
    if (pre_node) {
      pre_node.next = arr[count];
    }
    let node = arr[arr.length - 1 - count];
    arr[count].next = node;
    pre_node = node;
    count += 1;
  }
  if (!pre_node) {
    return;
  }
  if (count * 2 == arr.length) {
    pre_node.next = null;
  } else {
    let node = arr[(arr.length - 1) / 2];
    pre_node.next = node;
    node.next = null;
  }
}

export {};

File: Algorithm_Rust/examples/143. 重排链表.rs
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { val, next: None }
    }
}

impl Solution {
    pub fn reorder_list(head: &mut Option<Box<ListNode>>) {
        let mut arr: Vec<*mut ListNode> = Vec::new();
        let mut node = head.as_mut();

        while let Some(n) = node {
            arr.push(&mut **n as *mut _);
            node = n.next.as_mut();
        }

        let mut i = 0;
        let mut j = arr.len().saturating_sub(1);

        while i < j {
            unsafe {
                (*arr[i]).next = Some(Box::from_raw(arr[j]));
                j -= 1;
                if i < j {
                    (*arr[j + 1]).next = Some(Box::from_raw(arr[i + 1]));
                }
                i += 1;
            }
        }

        if let Some(&last) = arr.last() {
            unsafe {
                (*last).next = None;
            }
        }
    }
}

fn main() {}
